### ISSUE TITLE
Zone Raiders [1995] renders incorrectly in VS 64-bit builds

### ISSUE THREAD

**Issue by siana** [joncampbell123/dosbox-x#1950]
**Describe the bug**
With some dosbox-x executables, the game runs correctly; with others, it's unplayable due to completely garbled up software 3D rendering.

The issue occurs both in VGA mode and in VBE 640x400 and 640x480 modes, and is independent of the settings made in uvconfig (univbe). It also occurs in all processor emulation modes - simple, dynamic, normal...

However it does not occur with all dosbox-x executables.

**To Reproduce**
Launching the game and going in-game in Zone Raiders (1995)

Win32_builds\mingw - renders correctly
Win32_builds\mingw-lowend - renders correctly
Win32_builds\mingw-sdl2 - renders correctly
Win32_builds\mingw-sdldraw - renders correctly
Win32_builds\x86_Release - renders correctly
Win32_builds\x86_Release_SDL2 - renders correctly

Win64_builds\mingw - renders correctly
Win64_builds\mingw-lowend - renders correctly
Win64_builds\mingw-sdl2 - renders correctly
Win64_builds\mingw-sdldraw  - renders correctly
Win64_builds\x64_Release - RENDERING BROKEN
Win64_builds\x64_Release_SDL2  - RENDERING BROKEN

Linux build-debug (x86-64, built with gcc 9.3.0) - renders correctly

**Expected behavior**
All builds should display the same compatibility behaviour for the software that is emulated.

**Screenshots**
![image](https://user-images.githubusercontent.com/428656/96171543-5b9aa480-0f25-11eb-830b-95d86f55119f.png)

**Environment (please complete the following information):**
 - Windows 10 64-bit, release 20.04
 - for Linux version: Linux Mint 20 64-bit XFCE in Virtualbox
 - Dosbox-X release 0.83.6
 - Here's an example config file that i tested with. 
[dosbox-x.conf.txt](https://github.com/joncampbell123/dosbox-x/files/5387050/dosbox-x.conf.txt)

**Additional context**
Similar problem was reported against upstream Dosbox in their compatibility survey https://www.dosbox.com/comp_list.php?showID=897 by 'rlhelinski' however it affected, for them, the Linux build.

**Comment by Wengier** [comment]
Thanks for reporting! Yes, I can confirm it did occur in the release version 0.83.6 (64-bit Visual Studio builds). Could you please try the latest code (64-bit Visual Studio builds) instead and see if the issue occurs on your system? Thanks!

Download for 0.83.7-pre (64-bit Visual Studio SDL1 version): [dosbox-x-64-update.zip](https://github.com/Wengier/dosbox-x-wiki/blob/master/dosbox-x-64-update.zip?raw=true)

**Comment by rderooy** [comment]
Could this be caused by the FPU accuracy problem in the VS builds?

Could you try disabling the FPU emulation and see if the problem vanishes? Assuming the game can run without FPU...

Under [cpu] set fpu=false

**Comment by rderooy** [comment]
btw, this same issue was also raised by @DSN9990 in https://github.com/joncampbell123/dosbox-x/issues/1809

**Comment by siana** [comment]
With build provided by @Wengier, the game runs properly.

Unless i enable ''simple" core emulation, which breaks game rendering.

The game cannot be tested with fpu=false. According to documentation, it requires 486DX with FPU. When attempted, the game main thread locks up when going into the stage.

**Comment by siana** [comment]
On my Linux build as above (git tag dosbox-x-v0.83.6), if i go into include/dosbox.h and comment out #define HAS_LONG_DOUBLE and then re-make, the game is still operational as before. 

Also added an #error into src/fpu/fpu_instructions_longdouble.h to make sure it's no longer used. 

I can't quite see where else to deliberately break 80-bit floating point in Linux to validate whether the game is 80-bit floating point dependent.

**Comment by rderooy** [comment]
@siana yeah it was just a guess, seeing how the issue was limited to the VS builds, and the corruption was in the 3D rendered scene part only.

**Comment by Allofich** [comment]
> Additional context
Similar problem was reported against upstream Dosbox in their compatibility survey https://www.dosbox.com/comp_list.php?showID=897 by 'rlhelinski' however it affected, for them, the Linux build.

Yeah, you may have already seen it, but did you look at the linked screenshot? It looks just like your screenshot (just with the garbling happening mostly on the right, if that matters).

**Comment by Allofich** [comment]
Same person also posted about it on VOGONS.
https://www.vogons.org/viewtopic.php?t=67418

No resolution to the issue written there, and their last post was over a year ago so they may no longer be around, but if someone with a VOGONS account is interested it might be worthwhile to ask if they ever found out what the problem was.

**Comment by siana** [comment]
Nah, i'm not trying to get IP-banned on Vogons.

I'll just have to try build the Windows version in VS2019 and see if i can repro the issue; then dial down the optimisations and maybe it goes away.

Another possibility would be to build the Linux upstream and bisect it, if they fixed it. My computer is struggling, but i can manage.

At least for now, the dynarec is masking the issue and making the game operational, but something is not fixed. Which BTW newly integrated dynarec's performance is awesome!

**Comment by Allofich** [comment]
Wait, I hadn't read all the responses before writing my posts above, but I now see you wrote

> With build provided by @Wengier, the game runs properly.

which seems to indicate the problem is fixed? But then I get the impression from your latest post

> I'll just have to try build the Windows version in VS2019 and see if i can repro the issue; then dial down the optimisations and > maybe it goes away.
>
>At least for now, the dynarec is masking the issue and making the game operational, but something is not fixed. 

that there is still an issue?

**Comment by siana** [comment]
The issue is only no longer obvious because now for protected mode programs, the newly reintroduced dynamic CPU core is the default. On other builds except 64-bit VS, regardless of core type enabled, there is no issue.

With Wengier's provided build for [cpu]
core=auto - renders correctly
core=normal - RENDERING BROKEN
core=dynamic - renders correctly
core=full - RENDERING BROKEN
core=simple - RENDERING BROKEN

This means there is still an underlying correctness issue in the CPU interpreter, reeking of undefined behaviour. It might not practically matter for this specific game any longer - just as long as you're on x86 or x86_64 and you have the power of dynarec - but investigating it is still good for peace of mind, though probably not a major priority for you lot, and potentially not easy.

But have to at least narrow it down before deciding to give up :D
